src/server/internalServer.cpp

```diff
@@ -95,35 +95,174 @@ inline std::string normalizeRootUrl(std::string rootUrl)
+{
+  return ParameterizedMessage("raw-entry-not-found",
+                              {
+                                {"DATATYPE", dt},
+                                {"ENTRY", entry},
+                              }
+  );
+}
+
+ParameterizedMessage nonParameterizedMessage(const std::string& msgId)
+{
+  const ParameterizedMessage::Parameters noParams;
+  return ParameterizedMessage(msgId, noParams);
+}
+
+struct Error : public std::runtime_error {
+  explicit Error(const ParameterizedMessage& message)
+    : std::runtime_error("Error while handling request"),
+      _message(message)
+  {}
+
+  const ParameterizedMessage& message() const
+  {
+    return _message;
+  }
+
+  const ParameterizedMessage _message;
+};
+
 } // unnamed namespace
 
-SearchInfo::SearchInfo(const std::string& pattern)
-  : pattern(pattern),
-    geoQuery()
-{}
+Library::BookIdSet limitBooksNumber(Library::BookIdSet&& bookSet, unsigned int limit) {
+  if (limit == 0 || bookSet.size() <= limit) {
+    return std::move(bookSet);
+  }
+  auto it = bookSet.begin();
+  while(limit--) {
+    it++;
+  }
+  bookSet.erase(it, bookSet.cend());
+  return std::move(bookSet);
+}
 
-SearchInfo::SearchInfo(const std::string& pattern, GeoQuery geoQuery)
-  : pattern(pattern),
-    geoQuery(geoQuery)
-{}
+Library::BookIdSet InternalServer::selectBooks(const RequestContext& request, const std::string& prefix) const
+{
+  // Try old API
+  try {
+    auto bookName = request.get_argument("content");
+    try {
+      return {mp_nameMapper->getIdForName(bookName)};
+    } catch (const std::out_of_range&) {
+      throw Error(noSuchBookErrorMsg(bookName));
+    }
+  } catch(const std::out_of_range&) {
+    // We've catch the out_of_range of get_argument
+    // continue
+  }
+
+  // Does user directly gives us ids ?
+  try {
+    auto id_vec = request.get_arguments(prefix+"id");
+    if (id_vec.empty()) {
+      throw Error(noValueForArgMsg(prefix+"id"));
+    }
+    return limitBooksNumber(Library::BookIdSet(id_vec.begin(), id_vec.end()), m_multizimSearchLimit);
+  } catch(const std::out_of_range&) {}
 
-SearchInfo::SearchInfo(const RequestContext& request)
-  : pattern(request.get_optional_param<std::string>("pattern", "")),
-    geoQuery(),
-    bookName(request.get_optional_param<std::string>("content", ""))
+  // Use the names
+  try {
+    auto name_vec = request.get_arguments(prefix+"name");
+    if (name_vec.empty()) {
+      throw Error(noValueForArgMsg(prefix+"name"));
+    }
+    Library::BookIdSet bookIds;
+    for(const auto& bookName: name_vec) {
+      try {
+        bookIds.insert(mp_nameMapper->getIdForName(bookName));
+      } catch(const std::out_of_range&) {
+        throw Error(noSuchBookErrorMsg(bookName));
+      }
+    }
+    return limitBooksNumber(std::move(bookIds), m_multizimSearchLimit);
+  } catch(const std::out_of_range&) {}
+
+  // Check for filtering
+  Filter filter = get_search_filter(request, prefix+"query.");
```

Comment:
I think that  the word "filter" (instead of "query") is a better fit for the URL query parameter (so that the URL reads `.../search?books.filter.lang=eng&pattern=...`)